### ISSUE TITLE
VS Code Canvas — transient invalid source can blank Canvas and clear selection

### ISSUE THREAD

**Issue by sayosomi** [sayosomi/nuinuiCAD#295]
## Observed bug

Discovered during [SAY-88](https://linear.app/sayosomi/issue/SAY-88/canvas-selection-hidden-disabled-geometry-must-not-remain-selected) Manual E2E in the isolated VS Code Extension Development Host.

While editing valid nui4 source into a new valid state, an intermediate malformed / errorful source snapshot can sometimes make the Canvas go completely blank and clear the current Canvas selection. When the source becomes valid again, geometry returns, but the previous selection is not restored.

The trigger appeared intermittent because different intermediate source snapshots preserve or disturb statement identities differently.

## Investigation result

Baselines checked:

* latest remote `main`: `cf7793f4f9a4798982d3acf7ea2dc64f0bb151d9`
* [SAY-88](https://linear.app/sayosomi/issue/SAY-88/canvas-selection-hidden-disabled-geometry-must-not-remain-selected) observed branch / PR #299 head: `4a51a9aa0e8e83365bba31cc85e4d5d5af1a03e3`

PR #299 changes shared selection eligibility/reconciliation but does not change Canvas rendering, evaluation transport, or source-change publication. The underlying presentation race exists independently; [SAY-88](https://linear.app/sayosomi/issue/SAY-88/canvas-selection-hidden-disabled-geometry-must-not-remain-selected) makes the resulting selection invalidation more explicit.

### 1. Fatal compile handling is already last-good

`src/document/canonicalDocument.ts::compileCanonicalText()` already preserves the previous canonical `doc` when a compile is fatal. `cadDocumentStore.commitText()` then reconciles against that retained document.

Therefore a strictly fatal parser error by itself is **not** sufficient to explain the observed reset.

### 2. Some live authoring states are intentionally compileable despite error diagnostics

The DSL compiler deliberately excludes `missing-attribute-value` from its fatal document gate so authoring states such as an empty named value can still produce live source semantics / spans. Other partially-valid transitions can also temporarily change the parsed statement set or reconciled identities before the final valid edit is reached.

This behavior exists for editor authoring and should not be removed globally merely to stabilize Canvas.

### 3. Canvas can mix document revisions while Rust evaluation is stale

`src/geometry/useEvaluationEngine.ts` intentionally retains the previous async Rust evaluation while a new request is in flight. It marks that evaluation `isStale` and keeps its older `evaluationRevision`.

`src/components/DrawingCanvas.tsx`, however, passes that evaluation together with the **current** host-adapter `elements` into `useCanvasOverlayData()` / `renderCanvasGeometry()`.

`useCanvasOverlayData()` then derives visibility and metadata by joining `evaluation.computedGeometry` / `evaluation.effectiveVisibleElementIds` with the supplied current element IDs. If an intermediate source snapshot caused statement identity or element membership to change, the Canvas is therefore joining:

```text
old evaluation identity set
+ new/transient element identity set
```

rather than one revision-coherent presentation snapshot.

That mismatch can filter all previously computed geometry away and produce a blank Canvas even though the evaluation hook still holds usable last-rendered geometry.

### 4. Selection is reconciled against the transient element identities immediately

[SAY-88](https://linear.app/sayosomi/issue/SAY-88/canvas-selection-hidden-disabled-geometry-must-not-remain-selected)'s shared selection path correctly filters hidden/disabled and nonexistent IDs. On its branch, `cadDocumentStore.commitText()` publishes the compile result's elements and `reconcileSelectionWithElements()` normalizes the existing selection against those current identities.

If an errorful intermediate source snapshot changes/drops the selected statement identity, the previous selected ID is removed immediately. When the later valid source restores geometry, normal reconciliation does not reconstruct the user's old selection, so the selection loss persists.

This explains the two observed symptoms as one revision-stability problem:

* blank Canvas = stale evaluation joined with transient/current document identity inputs;
* permanent selection loss = transient/current identity set treated as authoritative for reconciliation.

### 5. `replaceTextDocument` is a separate reset path, not the normal typing path

The VS Code extension sends ordinary `onDidChangeTextDocument` edits as `commitText`. `replaceTextDocument` is used for initial authoritative publication/resync and intentionally resets selection/history. Do not fix this bug by weakening initial/resync reset semantics unless a focused test proves an unexpected resync is also involved.

### A. Revision-coherent Canvas presentation

1. Canvas must never combine an evaluation from one `compiledDocumentRevision` with element identity / activity / metadata inputs from another revision.
2. While Rust evaluation for a newer document is pending, either:
   * keep rendering the previous complete Canvas presentation snapshot as an atomic unit; or
   * use another equivalent revision-coherent representation.
3. If `EvaluationEngineState.isStale` is true, all document-owned inputs used to decide whether and how its geometry is drawn must correspond to that stale evaluation's source revision. Do not pair old computed geometry with the transient current element identity set.
4. Promote the newer Canvas presentation only when the evaluation/document pair is coherent. Do not introduce a flash-to-empty placeholder when a usable previous presentation exists.
5. Keep hit testing / drag mutation fail-closed against stale evaluation as today; stale presentation may remain visible, but it must not become authority for unsafe edits against a different canonical document revision.

### B. Selection stability during errorful editor authoring

 6. An editor-origin source snapshot that still has error-severity diagnostics must not permanently destroy an otherwise valid existing Canvas selection merely because its temporary parsed/reconciled identity set differs.
 7. Keep the current selection associated with the last interaction-safe / presentation-stable document until a source snapshot becomes authoritative for Canvas interaction again.
 8. Once a new **error-free** valid document becomes authoritative, normal selection reconciliation applies:
    * an intentionally deleted element is removed;
    * [SAY-88](https://linear.app/sayosomi/issue/SAY-88/canvas-selection-hidden-disabled-geometry-must-not-remain-selected) effective hidden/disabled pruning still applies;
    * surviving stable IDs stay selected.
 9. Do not restore a selection by matching names or source text heuristically. Preserve/reconcile stable identity only across explicit revision ownership.
10. Do not weaken [SAY-88](https://linear.app/sayosomi/issue/SAY-88/canvas-selection-hidden-disabled-geometry-must-not-remain-selected). A final valid `state: hidden` / `state: disabled` transition must still prune the target selection normally.

### C. Preserve live authoring/compiler semantics

11. Do not solve [SAY-102](https://linear.app/sayosomi/issue/SAY-102/vs-code-canvas-transient-invalid-source-can-blank-canvas-and-clear) by making every parser/compiler error globally fatal. In particular, do not remove the existing `missing-attribute-value` authoring carve-out solely for Canvas stability.
12. Source text, diagnostics, Completion/value-span authoring, and other live editor semantics may continue to observe the current source snapshot while Canvas presentation remains pinned to the last interaction-safe snapshot.
13. Keep `replaceTextDocument` reset semantics for genuine document load / authoritative resync unless a new deterministic resync-specific regression is demonstrated.
14. No separate VS Code parser/evaluator semantics. The fix must preserve the existing shared compiler/evaluator ownership and make the host presentation lifecycle revision-safe.

## Focused automated test contract

Add coverage for at least:

### Document / selection lifecycle

* valid selected document -> editor-origin errorful intermediate snapshot with changed/missing temporary element identity -> prior selection remains stable;
* subsequent error-free source with the same logical target identity keeps the selection;
* subsequent error-free source that intentionally deletes the target clears it;
* subsequent error-free `state: hidden` / `disabled` still prunes through [SAY-88](https://linear.app/sayosomi/issue/SAY-88/canvas-selection-hidden-disabled-geometry-must-not-remain-selected);
* unrelated valid selected members survive an errorful edit elsewhere;
* genuine `replaceTextDocument` document reset still clears selection/history as designed.

### Evaluation / Canvas revision pairing

* current document revision N+1 + stale evaluation revision N never feeds revision-N geometry through revision-N+1 element identity filtering;
* the previous renderable presentation remains non-empty while the newer Rust evaluation is pending;
* when the N+1 evaluation resolves, presentation advances atomically to N+1;
* stale geometry remains non-interactive for drag/mutation paths already guarded by evaluation currency;
* visibility / runtime-only filtering uses the same revision-owned element metadata as the evaluation being rendered.

Suggested owners include:

* `src/geometry/useEvaluationEngine.test.ts` / evaluation-engine focused coverage
* `src/components/DrawingCanvas.test.ts`
* `src/components/useCanvasOverlayData.test.tsx`
* `src/state/cadDocumentStore.test.ts`
* `src/vscode/VSCodeApp.test.tsx`

## Verification gates

Because this crosses shared document state, Canvas evaluation lifecycle, and the VS Code host adapter:

1. focused tests above
2. `npm run build`
3. `npm run build:vscode`
4. `npm run lint`
5. `npm test`

No Rust evaluator semantic change is intended. `npm run test:parity` and Rust tests are required only if implementation changes evaluation payload/semantics rather than presentation ownership.

## Manual E2E plan

Required: this bug depends on the asynchronous VS Code document/evaluation lifecycle.

Use the standard isolated VS Code Extension Development Host baseline from Project Context: fresh `--user-data-dir`, empty `--extensions-dir`, VS Code built-in completion disabled, task fixture generated outside the checkout, and `npm run build:vscode` before launch.

Fixture shape:

```nui
nui 4

point A = coordinate(
  x: 0,
  y: 0,
)

point B = coordinate(
  x: 60,
  y: 0,
)

line AB = segment(
  start: @A,
  end: @B,
)
```

Acceptance units:

1. Select `A` or `B` on Canvas. While editing Source, deliberately pass through errorful intermediate call/argument states and then restore a valid equivalent statement. Canvas must keep a usable last stable drawing and the valid prior selection; no blank flash followed by permanent selection loss.
2. Repeat with an errorful edit on an unrelated statement while another element is selected. Selection must remain unchanged.
3. Make an error-free intentional deletion of the selected element. Once the new revision is authoritative, selection must clear normally.
4. With [SAY-88](https://linear.app/sayosomi/issue/SAY-88/canvas-selection-hidden-disabled-geometry-must-not-remain-selected) semantics present, make an atomic valid `state: hidden` / `state: disabled` edit. Selection must prune normally; the transient-error guard must not suppress real final activity changes.
5. Confirm Canvas pointer/drag operations do not act on a stale visible presentation while a newer revision is pending.

The implementation Task must turn at least one observed transient source sequence into a deterministic automated regression fixture before completion; Manual E2E is the host-level lifecycle proof, not the only reproduction mechanism.

## Relation to [SAY-103](https://linear.app/sayosomi/issue/SAY-103/nui4-parser-blank-lines-inside-multiline-calls-incorrectly-produce) / [SAY-104](https://linear.app/sayosomi/issue/SAY-104/nui4-parser-named-arguments-incorrectly-require-whitespace-after-colon)

[SAY-103](https://linear.app/sayosomi/issue/SAY-103/nui4-parser-blank-lines-inside-multiline-calls-incorrectly-produce) and [SAY-104](https://linear.app/sayosomi/issue/SAY-104/nui4-parser-named-arguments-incorrectly-require-whitespace-after-colon) reduce common ways to enter unnecessary error states, but they do not own this lifecycle bug. [SAY-102](https://linear.app/sayosomi/issue/SAY-102/vs-code-canvas-transient-invalid-source-can-blank-canvas-and-clear) must remain safe for genuine future parse/type errors as well.

---
Original Linear issue: [SAY-102](https://linear.app/sayosomi/issue/SAY-102/vs-code-canvas-transient-invalid-source-can-blank-canvas-and-clear)
Linear status: Done
Linear priority: No priority
Linear labels: Passed, Ready, Bug
Linear project: None
Linear parent: None
Linear blocks: None
Linear blocked by: SAY-88
Linear related: SAY-103, SAY-104, SAY-105, SAY-113, SAY-117, SAY-118, SAY-155, SAY-161, SAY-21, SAY-61

**Comment by sayosomi** [comment]
Ready-contract freshness after SAY-88 merge: PR #299 is now merged into `main` at `b1e15796f9eda438600efa5fda6072ea37f7f5eb`. Treat the SAY-88 selection behavior described here as merged baseline behavior, not branch-only behavior. The SAY-102 implementation contract remains otherwise unchanged: transient errorful authoring must preserve revision-coherent Canvas presentation/selection, while final valid hidden/disabled transitions still prune through the merged SAY-88 invariant.

**Comment by sayosomi** [comment]
## SAY-102 resumed after publish — current baseline prepared

- Published prior changes: SAY-61 PR #561 merged to `main` at `33e66e8a6cf283686e222ab5fad2771888e15e3e`.
- SAY-102 implementation remains present on that current `main`; no implementation re-opened.
- SAY-88 prerequisite is Done.
- Automated verification on the exact baseline:
  - focused SAY-102 regression set: 9 files / 160 tests PASS
  - `npm run build` PASS
  - `npm run build:vscode` PASS
  - `npm run lint` PASS
  - full `npm test`: 512 passed, 8 skipped; 5,063 passed, 20 skipped
- E2E lane fixed to exact ref `33e66e8a6cf283686e222ab5fad2771888e15e3e`.
- Human host preparation succeeded and emitted `READY FOR HUMAN E2E`.
- Handoff: `/tmp/nuinui-SAY-102-human-e2e.env`
- Prepared fixture: `/tmp/nuinui-vscode-e2e.0TuN6x/nuinui-say-102.nui`
- CDP port: `9223`

No product GUI action has been counted yet; keep the Issue at `In Review + manual_e2e_only + Manual E2E: Ready to Run` until the human starts the units. Pending acceptance is Unit 1 and Unit 5; Units 2 / 3 / 4a / 4b prior PASS remains retained per the latest re-audit.

**Comment by sayosomi** [comment]
## Manual E2E started — human execution window open

The user explicitly resumed SAY-102 Manual E2E.

- Tested ref: `33e66e8a6cf283686e222ab5fad2771888e15e3e`
- Dedicated checkout: `/Users/yosomi/Code/nuinuiCAD-e2e` (detached, clean, exact ref)
- E2E marker: present for SAY-102
- Isolated host/CDP: alive on port `9223`
- Handoff: `/tmp/nuinui-SAY-102-human-e2e.env`

Codex will not perform or score GUI product actions. Human should execute the pending Unit 1 and Unit 5 oracles against the already-prepared host, preserving the exact Unit 1 atomic editor operation and the Unit 5 stale-drag fail-closed oracle. Keep the Issue as `In Review + manual_e2e_only` with Manual E2E = Running while the human run is active.

**Comment by sayosomi** [comment]
## E2E procedure correction — VS Code surface has no Inspector

The previous handoff incorrectly instructed Human Unit 5 to open the shared-app right-side Inspector. That is not present in the current VS Code Extension Host surface.

Verified on tested ref `33e66e8a6cf283686e222ab5fad2771888e15e3e`:
- `src/vscode/VSCodeApp.tsx` mounts only `VSCodeDrawingCanvas` inside the webview.
- `RightPanel` / `InspectorPanel` are not mounted by the VS Code app.
- `Rust評価中 / stale` is therefore not a visible VS Code UI label.

Correction:
- Unit 1 can run using Source + Canvas + Problems only; no Inspector step.
- Unit 5 must not be scored from timing guesses or visual Canvas state alone. Under the declared oracle, it is BLOCKED unless an objective read-only observation channel exposes `evaluationStatus` / `isStale` during the drag attempt. The prepared host has the developer observation bridge enabled, but it is not a visible Inspector.
- No product FAIL is inferred from this UI mismatch; the prior procedure was stale and invalid.

**Comment by sayosomi** [comment]
## Human E2E Unit 1 result — PASS

User-reported result: Unit 1 PASS on the prepared current tested ref `33e66e8a6cf283686e222ab5fad2771888e15e3e`.

Unit 1 oracle:
- transient missing-`end` error state was exercised;
- last stable Canvas remained usable;
- A selection remained preserved through the error and recovery.

Unit 5 remains unscored because the current VS Code Canvas surface has no visible Inspector/stale indicator. Under the declared binary oracle, do not infer PASS from timing alone; it remains BLOCKED until an objective observation channel is available.

**Comment by sayosomi** [comment]
## Human E2E Unit 5 — BLOCKED by infeasible observation procedure

User reported that synchronizing a manual Canvas drag with the short `isStale=true` interval is not feasible.

Classification:
- Unit 1: PASS retained.
- Unit 5: BLOCKED, not product FAIL.
- Cause: the current VS Code UI has no visible stale indicator, and the proposed external watcher requires human timing coordination that is not practically executable.
- No source or product regression is inferred.

Do not ask the user to retry the same timing procedure. Unit 5 requires a contract/procedure correction: either a deterministic host-controlled stale interval / visible VS Code status, or explicit removal/reclassification of this Manual E2E unit after reviewing existing automated stale-mutation coverage.

**Comment by sayosomi** [comment]
## Sol contract re-audit — Unit 5 reclassified to automated verification; Manual E2E PASS

Re-audited the active SAY-102 contract on exact current main `33e66e8a6cf283686e222ab5fad2771888e15e3e`.

### Unit 5 classification correction

The required behavior remains: stale Canvas presentation must not become mutation authority. The Human timing procedure is removed from Manual E2E because it is not practically executable and duplicates direct automated production-path coverage.

Existing coverage protects the exact contract:
- `src/components/DrawingCanvas.test.ts`: pointerdown against stale/non-current evaluation is deferred; held drag resolves only after matching current evaluation.
- `src/vscode/VSCodeDrawingCanvas.test.tsx`: stale evaluation does not flush queued drag preview; latest action proceeds only after current settlement.
- `src/vscode/vscodeDragPreviewScheduler.test.ts`: stale ready/failed evaluations are not accepted as settlement.
- Production wiring passes `useEvaluationEngine` state through `VSCodeApp -> VSCodeDrawingCanvas -> DrawingCanvas`; the tests exercise the production components/scheduler rather than a disconnected helper.

Per current Manual E2E policy, behavior sufficiently proved by deterministic automated tests is not a required Manual E2E unit. No product or source change is needed.

### Final result

- Unit 1: Human production-host PASS on current exact ref.
- Units 2 / 3 / 4a / 4b: prior PASS retained after freshness review.
- Unit 5: automated verification PASS; removed from Human Manual E2E.
- Focused SAY-102 suite: 9 files / 160 tests PASS.
- Build, VS Code build, lint: PASS.
- Full test suite: 512 files / 5,063 tests PASS; skips unchanged.
- Blocking contract review: PASS, no blocking finding.

Aggregate Manual E2E: Passed. SAY-102 has no remaining completion gate.
